### ISSUE TITLE
Optimise storage iteration for the usual case (only one iteration request active)

### ISSUE THREAD

**Issue by gavin-norman-sociomantic** [sociomantic-tsunami/dhtnode#46]
As each TokyoCabinet instance only has a single iterator, `StorageEngine.getNextKey` currently always restarts the iteration based on the last key that was returned by the iterator (https://github.com/sociomantic-tsunami/dhtnode/blob/master/src/dhtnode/storage/StorageEngine.d#L616-L617). I think this (restarting an iteration at an arbitrary key) is a relatively expensive operation.

Looking at the node request stats, I see that most of the time only a single GetAll request is active, so this constant restarting of the iteration is probably not necessary. It only needs to be restarted it the iterator for that instance has been moved by another request.

This could provide a pretty significant CPU performance improvement to GetAll.

**Comment by gavin-norman-sociomantic** [comment]
Yep, a local test of doing a single GetAll over a small channel shows the following:
* Microseconds spent iterating, before optimisation: ~202,000.
* Microseconds spent iterating, after optimisation: ~126,000.

**Comment by nemanja-boric-sociomantic** [comment]
What happens if you make the channel size 10x more? I wonder if these 100us are just fixed price for restarting the iterator.

**Comment by gavin-norman-sociomantic** [comment]
100**K** microseconds. But yeah, I'll try a bigger channel too.

**Comment by nemanja-boric-sociomantic** [comment]
Why didn't you then express it in `ms` :-D . Sorry, my bad. Yeah, 100ms is something worth fighting for !

**Comment by gavin-norman-sociomantic** [comment]
It's just my way of measuring CPU usage. (Total time spent calling `iterator.next()`.)

**Comment by gavin-norman-sociomantic** [comment]
Because the `StopWatch` works in microseconds ;)

**Comment by gavin-norman-sociomantic** [comment]
Actually, my test code was wrong :( The real (with a much bigger channel) results are:
* Microseconds spent iterating, before optimisation: ~1,175,183.
* Microseconds spent iterating, after optimisation: ~970,000.

So, it makes a little difference, but it seems that restarting the iteration isn't as big a deal as I suspected.

**Comment by gavin-norman-sociomantic** [comment]
>I wonder if these 100us are just fixed price for restarting the iterator.

Maybe you didn't get that the old code is restarting the iterator _after every iteration step_.

**Comment by nemanja-boric-sociomantic** [comment]
Ah, no, I didn't get this.

**Comment by gavin-norman-sociomantic** [comment]
That's why I've always suspected this might be horribly inefficient. (And we do see very high CPU usage when a dhtnode is handling GetAll.)

**Comment by gavin-norman-sociomantic** [comment]
Testing with `perf stat` also shows that this doesn't make much difference. Oh well. It's good to have finally checked!
